compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py

```diff
@@ -58,13 +61,56 @@ def main(args):
         print(m.get_asm(assume_verified=not args.no_verify))
 
 
-def load_onnx_model(file_path: Path) -> onnx.ModelProto:
-    raw_model = onnx.load(file_path)
-    inferred_model = onnx.shape_inference.infer_shapes(raw_model)
-    return inferred_model
+def load_onnx_model(args: argparse.Namespace) -> onnx.ModelProto:
+    input_dir = os.path.dirname(os.path.abspath(args.input_file))
 
-
-def parse_arguments(argv=None):
+    # Load the model, with possible external data coming from the default
+    # location, or the location specified on the command line.
+    if args.data_dir is None:
+        raw_model = onnx.load(args.input_file)
+    else:
+        raw_model = onnx.load(args.input_file, load_external_data=False)
+        onnx.load_external_data_for_model(raw_model, args.data_dir)
+
+    # Do shape inference two ways.  First, attempt in-memory to avoid redundant
+    # loading and the need for writing a temporary file somewhere.  If that
+    # fails, typically because of the 2 GB protobuf size limit, try again via
+    # files.  See
+    # https://onnx.ai/onnx/repo-docs/PythonAPIOverview.html#shape-inference-a-large-onnx-model-2gb
+    # for details about the file-based technique.
+
+    # Run the checker to test whether the file is above the threshold for
+    # in-memory shape inference.  If not, go ahead and do the shape inference.
+    try:
+        onnx.checker.check_model(raw_model)
+        inferred_model = onnx.shape_inference.infer_shapes(
+            raw_model, data_prop=args.data_prop
+        )
+        return inferred_model
+    except ValueError:
+        pass
+
+    # Model is too big for in-memory inference: do file-based shape inference
+    # to a temp file.
+    # Make a temp dir for all the temp files we'll be generating as a side
+    # effect of infering shapes. For now, the only file is a new .onnx holding
+    # the revised model with shapes.
+    with tempfile.TemporaryDirectory(dir=input_dir) as temp_dir_name:
+        temp_dir_path = Path(temp_dir_name)
+        temp_inferred_file = temp_dir_path / "temp-inferred.onnx"
+        onnx.shape_inference.infer_shapes_path(
+            args.input_file, temp_inferred_file, data_prop=args.data_prop
+        )
+
+        # Load the temp file and the external data.
+        inferred_model = onnx.load(temp_inferred_file, load_external_data=False)
+        data_dir = Path(input_dir if args.temp_dir is None else args.data_dir)
+        onnx.load_external_data_for_model(inferred_model, data_dir)
```

Comment:
This `data_dir` code looks wrong. The temp_dir arg is no longer used, so what is this actually trying to do?

I'm not really following the logic here. New tests cases would help: https://github.com/iree-org/iree/blob/main/compiler/bindings/python/test/tools/import_onnx_test.py.

(note that those aren't running on CI: https://github.com/iree-org/iree/issues/16624)

Reply:
I think that this is the typo from whoever wrote the torch-mlir importer. They probably meant args.data_dir. I don't see how these two could be related

Reply:
Ah I see, can add CI for this in a follow up PR. I've been locally testing that it works. Also, for all e2eshark testing we don't use the shape inference with a file. Might have to figure out which cases require this. Might get more models passing

Reply:
Yeah no need to get the CI running as part of this. I'm just hesitant to accept clearly buggy code from upstream blindly :P. If the shape inference with a file path hasn't been exercised yet, we could also remove it until needed... smaller surface area that way.

Reply:
Okay, will just remove the shape inference with a file path then. I haven't seen it used anywhere

Reply:
Or actually, I'll leave it in case it is required somewhere. Logic seems to be correct from what I can tell, but I'll try to get some tests for all these cases.